File: src/BiAi/Services/ImageProcessor.cs
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using BiAi.Models;
using BiAi.Models.Config;
using LanguageExt;
using LanguageExt.Common;
using Microsoft.Extensions.Logging;

namespace BiAi.Services;

public class ImageProcessor : IImageProcessor
{
    private readonly IDeepStackService _deepStackService;
    private readonly ILogger<ImageProcessor> _logger;
    private readonly ITelegramService _telegramService;
    private readonly ITriggerService _triggerService;

    public ImageProcessor(
        ILogger<ImageProcessor> logger,
        IDeepStackService deepStackService,
        ITelegramService telegramService,
        ITriggerService triggerService)
    {
        _logger = logger;
        _deepStackService = deepStackService;
        _telegramService = telegramService;
        _triggerService = triggerService;
    }

    public async Task ProcessImageAsync(CameraConfig camera, Image image, CancellationToken cancellationToken)
    {
        _logger.LogDebug(
            "Image at {imagePath} was created and matched to camera {camera}",
            image.FullPath,
            camera.Name);

        if (camera.TelegramTrigger.IsInCooldown(image) && camera.Triggers.All(t => t.IsInCooldown(image)))
        {
            _logger.LogDebug(
                "All triggers are in cooldown for camera {camera}, ignoring image at {imagePath}",
                camera.Name,
                image.FullPath
            );
        }
        else
        {
            Either<Error, DeepStackResponse> response = await _deepStackService.DetectAsync(
                image.FullPath,
                cancellationToken);
            await response
                .Right(async r => await ProcessDeepStackResponseAsync(camera, r, image, cancellationToken))
                .Left(async error => await Task.Run(() => _logger.LogWarning(error.Message), cancellationToken));
        }
    }

    private async Task ProcessDeepStackResponseAsync(
        CameraConfig camera,
        DeepStackResponse response,
        Image image,
        CancellationToken cancellationToken)
    {
        if (response.Success)
        {
            Seq<DeepStackObject> relevantObjects = GetRelevantObjects(camera, response);
            if (!relevantObjects.IsEmpty)
            {
                _logger.LogDebug(
                    "Detected relevant objects {objects}",
                    relevantObjects.Map(o => new { o.Label, o.Confidence }));

                if (camera.TelegramTrigger != null)
                {
                    await _telegramService.ProcessTriggerAsync(camera.TelegramTrigger, image, cancellationToken);
                }

                await _triggerService.ProcessTriggersAsync(camera, image, cancellationToken);
            }
        }
        else
        {
            _logger.LogWarning("DeepStack did not detect any objects");
        }
    }

    private static Seq<DeepStackObject> GetRelevantObjects(CameraConfig camera, DeepStackResponse response) =>
        response.Predictions.ToSeq()
            .Filter(
                p => camera.RelevantObjects.Contains(p.Label)
                     && p.Confidence * 100 >= camera.LowerCertainty
                     && p.Confidence * 100 <= camera.UpperCertainty);
}

File: biai/src/biai/ImageProcessor.scala
package biai

import better.files.File
import cats.data.NonEmptyList
import cats.effect.IO
import cats.implicits._

trait ImageProcessor {
  def processImage(file: File): IO[Unit]
}

class ImageProcessorImpl(config: AppConfig, deepStackService: DeepStackService, logger: Logger) extends ImageProcessor {
  override def processImage(file: File): IO[Unit] =
    for {
      image <- Image.load(file)
      camera <- config.cameras.find(c => c.enabled && c.name == image.cameraName).liftTo[IO](CameraNotFound(image.file))
      _ <- logger.log(s"Image at ${image.file.canonicalPath} was created and matched to camera ${camera.name}")
      // TODO: check cooldown(s)
      response <- deepStackService.detectObjects(image)
      objects <- getRelevantObjects(camera, response)
      _ <- logRelevantObjects(objects)
      // TODO: process telegram trigger
      // TODO: process generic trigger(s)
    } yield ()

  private def getRelevantObjects(camera: CameraConfig, response: DeepStackResponse): IO[NonEmptyList[DeepStackObject]] =
    if (response.success)
      response.predictions
        .filter { o =>
          val isRelevant = camera.relevantObjects.contains(o.label)
          val confidence = o.confidence * 100
          isRelevant && confidence >= camera.minConfidence && confidence <= camera.maxConfidence
        }
        .toNel
        .liftTo[IO](NoRelevantObjectsDetected)
    else IO.raiseError(NoObjectsDetected)

  private def logRelevantObjects(objects: NonEmptyList[DeepStackObject]): IO[Unit] = {
    logger.log(s"Detected relevant objects ${objects.map(o => (o.label, o.confidence * 100))}")
  }
}
